lib/hyacinth/adapters/external_identifier_adapter/datacite.rb

```diff
@@ -17,96 +17,81 @@ def handles?(id)
-          datacite_metadata = Datacite::MetadataBuilder.new hyacinth_metadata
-          # ApiSession#modify_identifier returns true if the response from the EZID server indicated
-          # success, else it returns false
-          opts = { datacite: datacite_metadata.datacite_xml, _status: IDENTIFIER_STATUS[:public] }
-          opts[:_target] = location_uri unless location_uri.blank?
-          api_session.modify_identifier(id, opts)
-        end
-
-        # Following method will make a request to the EZID server to update the target URL associated
-        # with the EZID DOI entry on the server.
-        # (See _target in http://ezid.cdlib.org/doc/apidoc.html#internal-metadata)
-        # To delete the target URL stored in the DOI server, use an empty string as the argument.
-        # Uses the modify identifier API call, returns true if metadata update was successful
-        # if not, returns false
-        def update_doi_target_url(doi, target_url)
-          return false if doi.nil?
-          return false if target_url.nil?
-          # ApiSession#modify_identifier returns true if the response from the EZID server indicated
-          # success, else it returns false
-          api_session.modify_identifier(doi, _target: target_url, _status: IDENTIFIER_STATUS[:public])
+        # @return [Hash] hash containing the publishing info ready to be JSONified
+        def as_datacite_properties(digital_object, target_url)
+          HyacinthMetadata.as_datacite_properties(digital_object, target_url)
         end
 
-        def tombstone_impl(id)
-          return false if id.nil?
-          # ApiSession#modify_identifier returns true if the response from the EZID server indicated
-          # success, else it returns false
-          api_session.modify_identifier(id, _status: IDENTIFIER_STATUS[:unavailable])
+        # Returns true if an identifier exists in the external management system
+        def exists?(doi)
+          resp = rest_api.get_dois(doi)
+          rest_api.parse_doi_from_api_response_body(resp.body).present?
         end
 
-        private
-
-          # setup EZID API info: credentials, url, etc.
-          def api_session
-            Thread.current[:datacite_ezid_api_session] ||= Datacite::EzidSession.new(DATACITE[:user], DATACITE[:password])
-          end
-
-          def http_error_message(response)
-            return "" if response.nil?
-            "HTTP status code: #{response.http_status_code}\n" \
-            "HTTP server message: #{response.http_server_message}\n" \
-            "response body: #{response.body}"
+        # @param id [String]
+        # @param digital_object [DigitalObject]
+        # @param location_uri [String]
+        # @param doi_state [Symbol] doi_state can be set to one of the following: :draft, :findable, :registered
+        # @return [Boolean] true if update was successful, false otherwise
+        def update_impl(doi, digital_object:, location_uri:, doi_state: nil, **_args)
```

Comment:
The undocumented and unused `**_args` is throwing me off a bit here.  Can it be refactored away?

Reply:
I don't think so - the abstract signature I am suggesting is `(identifer, args_hash)`. The keyword args let us be a little stronger about what must be in the hash; the `**_args` lets the hash have additional keys (otherwise Ruby rejects calls with additional keys). In the JS parts of the app, we call the analog to this `rest`. Would it help to use that name in the Ruby parts of the app, too?

Reply:
Without doing further customization in rubocul, `**_rest` is as close as we can get.

Reply:
Another solution may be to update the abstract and memory adapters to match this signature.  The `doi`, `digital_object` and `doi_state` seem like they could be applicable to any external identifier adapter, but the **_args aren't used here.